source/common/filter/tcp_proxy.h

```diff
@@ -83,13 +90,18 @@ class TcpProxyConfig {
     std::string cluster_name_;
   };
 
-  static TcpProxyStats generateStats(const std::string& name, Stats::Scope& scope);
+  static TcpProxyStats generateStats(Stats::Scope& scope);
+
+  // Hold a Scope for the lifetime of the configuration because connections in
+  // the TcpProxyUpstreamDrainManager can live longer than the listener.
+  Stats::ScopePtr stats_scope_;
 
   std::vector<Route> routes_;
   const TcpProxyStats stats_;
   std::vector<AccessLog::InstanceSharedPtr> access_logs_;
   const uint32_t max_connect_attempts_;
   Optional<std::chrono::milliseconds> idle_timeout_;
+  ThreadLocal::SlotPtr upstream_drain_manager_slot_;
```

Comment:
This is really tricky, but I think there are still lifetime issues here. Consider the following sequence of events:
1) There are connections draining on worker threads.
2) Listener is deleted via LDS.
3) No connections, immediate drain/delete.
4) Main thread releases shared ref on config.
5) Drainer eventually times out / drains on worker.
6) Worker releases ref on config and config is deleted.

In the current code (https://github.com/envoyproxy/envoy/blob/master/source/common/thread_local/thread_local_impl.cc) We assume that slots always get created/deleted on the main thread. This will break (and assert if a test covered this).

How to fix this: The easiest way IMO is to make this effectively two staged. Have a config object which is used only on the main thread. Within the config object, have a struct for things that need to be accessed on all threads, and pass this via shared pointer to the drain manager when a connection is being drained. This would include stats (not sure what else), but *not* the TLS slot. Thus, the TLS slot will get blown away on the main thread.

This solution has the benefit that when the listener goes away via LDS, the slot will go away, then the slot deletion will be posted to all threads, where the manager will go away, and force the connections to close, which is arguably correct to do.

Let me know if ^ makes sense or not. Happy to talk in Slack also.

Reply:
That makes sense; I was for some reason thinking that there was only one config, created on the main thread, that is shared.  

Step 1 is that I'm going to create an integration test for this (now that @alyssawilk has very helpfully moved the tests to v2). Then I'll fix this and make sure it can pass an integration test without asserting.

Reply:
Ok, I'm getting back to this and re-read your comment; I didn't read it carefully enough the first time.

What would hold the reference to the main-thread-only-config?  I agree it would be ideal for the manager to go away if the listener goes away.

However, there's an alternate, much-simpler implementation: allocate the slot on the main thread on first use, and share it among all TcpProxyConfigs, as a static/global.  There's no reason each TcpProxy needs it's own manager.  I think I'd prefer the simpler implementation, even though it loses the benefit of connections getting closed when the LDS listener goes away; I think the idle-timeout makes that case good-enough.

Reply:
I would rather not start using static/global for this type of stuff. It makes proper shutdown very hard/impossible without doing a bunch more work. I don't think it's that hard to do it with a slight variation of the code you have.

Take a look at my ongoing gRPC access log PR: https://github.com/envoyproxy/envoy/commit/8345c1f6a891f1cad56331231e92c31ab4eb889a#diff-94ea476ae8765d98909a0305244e49f3R33

I think you want to basically do the same thing. Essentially you have a top level config which holds the slot. The config owns some shared state which is owned both the config and all of the thread locals. Then the slot gets destroyed on the main thread, the thread destructions go out to the workers, and then last worker to free the shared state frees it.